File: prob1.java
// Time Complexity : O(n)
// Space Complexity : O(1)


class Solution {
    public boolean isRobotBounded(String instructions) {
        int x = 0, y = 0;
        char dir = 'N';
        for(char ch : instructions.toCharArray()){
            if(ch == 'G'){
                switch(dir){
                    case 'N':   y++; break;
                    case 'E':   x++; break;
                    case 'W':   x--; break;
                    case 'S':   y--; break;          
                }
            }
            else if( ch == 'L'){
                switch(dir){
                    case 'N':   dir = 'W'; break;
                    case 'E':   dir = 'N'; break;
                    case 'W':   dir = 'S'; break;
                    case 'S':   dir = 'E'; break;       
                } 
            }
            else{
                switch(dir){
                    case 'N':   dir = 'E'; break;
                    case 'E':   dir = 'S'; break;
                    case 'W':   dir = 'N'; break;
                    case 'S':   dir = 'W'; break;          
                }
            }
        }

        if(x == 0 && y == 0) return true;

        if(dir == 'N') return false;
        
        return true;
    }

    
}

File: prob1.py
# Time Complexity : O(n)
# Space Complexity :O(1)
# Passed on Leetcode: yes

class Solution:
    def isRobotBounded(self, instructions: str) -> bool:

        dirs = [[0,1],[1,0],[0,-1], [-1,0]] #clockwise N,E,S,W
        idx = 0
        x, y = 0, 0
        for i in instructions:
            if i == 'R':
                idx = (idx+1)%4
            
            elif i == 'L':
                idx = (idx+3)%4
            
            else:
                x = dirs[idx][0] + x
                y = dirs[idx][1] + y
        if x == 0 and y == 0:
            return True
        if dirs[idx][1] != 1:
            return True
        return False